wazo_test_helpers/asset_launching_test_case.py

```diff
@@ -59,38 +66,68 @@ def __init__(self, stdout, stderr, return_code):
         self.return_code = return_code
 
 
-class AssetLaunchingTestCase(unittest.TestCase):
-    """
-    Subclasses of this class MUST have the following fields:
+R = TypeVar("R", covariant=True)
 
-    * service: The name of the service under test in the docker-compose.yml file
-    * asset: The name of the asset to run
-    * assets_root: The location of the assets
-    """
 
-    cur_dir = None
-    log_dir = None
+class CachedClassProperty(Generic[R]):
+    cls: type
 
-    @staticmethod
-    def is_managing_containers():
-        return os.environ.get('TEST_DOCKER', 'manage') != 'ignore'
+    def __init__(self, func: Callable[..., R]) -> None:
+        self.func = func
 
-    @classmethod
-    def setUpClass(cls):
-        if cls.is_managing_containers():
-            cls.launch_service_with_asset()
-        else:
-            logger.debug('Container management disabled.')
+    def __set_name__(self, owner: type, name: str) -> None:
+        self.cls = owner
+        self.private_name = f'_{name}'
 
-    @classmethod
-    def tearDownClass(cls):
-        if cls.is_managing_containers():
-            cls.stop_service_with_asset()
-        else:
-            logger.debug('Container management disabled.')
+    def __get__(self, obj: Any | None = None, obj_type: type | None = None) -> R:
+        if obj is not None:
+            raise RuntimeError('Must not be called on instance')
+        if not hasattr(self.cls, self.private_name):
+            value = self.func(self.cls)
+            setattr(self.cls, self.private_name, value)
+            return value
+        return getattr(self.cls, self.private_name)
+
+    def __del__(self, obj: Any | None = None) -> None:
+        if obj is not None:
+            raise RuntimeError('Must not be called on instance')
```

Comment:
Out of curiosity, when/why this method is called with an `obj`?

Reply:
These methods would be called with an `obj` when called on an instance. Normally this shouldn't happen currently everything is used at the class level for some reason. I just though it would be a good idea to make the error clear instead of it just failing with some unclear error.